src/librustc_mir/interpret/operand.rs

```diff
@@ -682,16 +687,31 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                         (dataful_variant.as_u32() as u128, dataful_variant)
                     },
                     Ok(raw_discr) => {
-                        let adjusted_discr = raw_discr.wrapping_sub(niche_start)
-                            .wrapping_add(variants_start);
-                        if variants_start <= adjusted_discr && adjusted_discr <= variants_end {
-                            let index = adjusted_discr as usize;
-                            assert_eq!(index as u128, adjusted_discr);
+                        // We need to use machine arithmetic to get the relative variant idx.
+                        let discr_layout = self.layout_of(discr_layout.value.to_int_ty(*self.tcx))?;
+                        let discr_val = ImmTy::from_uint(raw_discr, discr_layout);
+                        let niche_start_val = ImmTy::from_uint(niche_start, discr_layout);
+                        let variant_index_relative_val = self.binary_op(
+                            mir::BinOp::Sub,
+                            discr_val,
+                            niche_start_val,
+                        )?;
+                        let variant_index_relative = variant_index_relative_val
+                            .to_scalar()?
+                            .assert_bits(discr_val.layout.size);
+                        // Then computing the absolute variant idx should not overflow any more.
+                        let variant_index = variants_start
+                            .checked_add(variant_index_relative)
+                            .expect("oveflow computing absolute variant idx");
+                        // Check if this is in the range that indicates an actual discriminant.
+                        if variants_start <= variant_index && variant_index <= variants_end {
```

Comment:
Oh wait you can just compare `variant_index_relative` against `variants_end - variants_start`, and then you can move the addition inside the "niche" side of the `if` (and it can be just `+` since it can't possibly overflow given the condition).

Reply:
So the if should change to `variant_index_relative <= variants_end - variants_start` you say?

What is the "niche side of the if"?